providers/implementations/keymgmt/build.info

```diff
@@ -14,7 +13,10 @@ IF[{- !$disabled{dsa} -}]
   SOURCE[../../libnonfips.a]=dsa_kmgmt.c
 ENDIF
 IF[{- !$disabled{ec} -}]
-  SOURCE[$EC_GOAL]=ec_kmgmt.c
+  SOURCE[../../libfips.a]=ec_kmgmt.c
+  DEFINE[../../libfips.a]=OPENSSL_NO_SM2
```

Comment:
You should consider `#ifndef FIPS_MODULE` in ec_kmgmt.c. That way, there's no need for this define

Reply:
I did: https://github.com/openssl/openssl/pull/13359#discussion_r527063051

Do you think that would be better than using a define here?

Reply:
I was playing with this as an alternative

```diff
diff --git a/providers/implementations/keymgmt/build.info b/providers/implementations/keymgmt/build.info
index b05501998f..f434a720bc 100644
--- a/providers/implementations/keymgmt/build.info
+++ b/providers/implementations/keymgmt/build.info
@@ -14,8 +14,6 @@ IF[{- !$disabled{dsa} -}]
 ENDIF
 IF[{- !$disabled{ec} -}]
   SOURCE[../../libfips.a]=ec_kmgmt.c
-  DEFINE[../../libfips.a]=OPENSSL_NO_SM2
-
   SOURCE[../../libnonfips.a]=ec_kmgmt.c
 ENDIF
 
diff --git a/providers/implementations/keymgmt/ec_kmgmt.c b/providers/implementations/keymgmt/ec_kmgmt.c
index cc88bb9278..e71c3344a6 100644
--- a/providers/implementations/keymgmt/ec_kmgmt.c
+++ b/providers/implementations/keymgmt/ec_kmgmt.c
@@ -13,6 +13,13 @@
  */
 #include "internal/deprecated.h"
 
+#if defined(FIPS_MODULE) && !defined(OPENSSL_NO_SM2)
+# define OPENSSL_NO_SM2
+#endif
+#if defined(FIPS_MODULE) && !defined(OPENSSL_NO_EC2M)
+# define OPENSSL_NO_EC2M
+#endif
+
 #include "e_os.h" /* strcasecmp */
 #include <string.h>
 #include <openssl/core_dispatch.h>
@@ -27,7 +34,6 @@
 #include "prov/providercommonerr.h"
 #include "prov/provider_ctx.h"
 #include "internal/param_build_set.h"
-#include "crypto/sm2.h"
 
 static OSSL_FUNC_keymgmt_new_fn ec_newdata;
 static OSSL_FUNC_keymgmt_gen_init_fn ec_gen_init;
@@ -50,7 +56,10 @@ static OSSL_FUNC_keymgmt_import_types_fn ec_import_types;
 static OSSL_FUNC_keymgmt_export_fn ec_export;
 static OSSL_FUNC_keymgmt_export_types_fn ec_export_types;
 static OSSL_FUNC_keymgmt_query_operation_name_fn ec_query_operation_name;
+
 #ifndef OPENSSL_NO_SM2
+# include "crypto/sm2.h"
+
 static OSSL_FUNC_keymgmt_gen_fn sm2_gen;
 static OSSL_FUNC_keymgmt_get_params_fn sm2_get_params;
 static OSSL_FUNC_keymgmt_gettable_params_fn sm2_gettable_params;
```

but somehow it felt better to define `OPENSSL_NO_SM2` (and `OPENSSL_NO_EC2M`, but maybe in a separate PR, as it is unrelated here) at the `build.info` rather than at the top of the file.
I don't have a strong preference either way.

Reply:
I would rather have something that is like this:

```
#ifndef FIPS_MODULE

/* ... SM2 stuff ... */

#endif
```

Where the SM2 stuff inside includes the `#ifndef OPENSSL_SM2` clauses.

Reply:
This is for detection reasons...

Reply:
I am not particularly fun of the proliferation of ifndef (and nested ones) in the code, but I will comply.

I am just curious, could you please elaborate on the benefits of the multiple nested ifdefs vs my proposal of conditionally declaring OPENSSL_NO_whatever at the incipit of the file?
(really, this doesn't want to be polemic, I just desire to understand the motivation better)

Reply:
It's to make detection of what is FIPS and what isn't FIPS source.  We have a plan and I have a long overdue PR that will parse through source and make a checksum over all the FIPS source, and therefore need an easy enough way to skip over non-FIPS source.

This is the same problem as, for example, `#ifndef OPENSSL_NO_XXX` in headers to detect specific features, something `util/mknum.pl` does.  We shouldn't have to write a complete AI to do this stuff, and that requires a bit of compliance when writing the source.

Reply:
Thanks for the explanation!

See 506418b62862a4cd5b90bf987ee11a203c525295 for the fixup.